Library/Source/ticketed_collection.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <map>
+
+namespace Babylon
+{
+    // NOTE: This type is not thread-safe.
+    template<typename T>
+    class ticketed_collection
+    {
+        using MapT = std::map<bool*, T>;
+
+    public:
+        class ticket
+        {
+        public:
+            ~ticket()
+            {
+                if (m_isCollectionStillAlive != nullptr && *m_isCollectionStillAlive)
+                {
+                    m_collection.erase(m_isCollectionStillAlive);
+                }
+
+                delete m_isCollectionStillAlive;
+            }
+
+            ticket(const ticket&) = delete;
+            
+            ticket(ticket&& other)
+                : m_isCollectionStillAlive{ other.m_isCollectionStillAlive }
+                , m_collection{ other.m_collection }
+            {
+                other.m_isCollectionStillAlive = nullptr;
+            }
+
+        private:
+            friend class ticketed_collection;
+
+            ticket(T&& value, MapT& collection)
+                : m_isCollectionStillAlive{ new bool{true} }
+                , m_collection{ collection }
+            {
+                m_collection[m_isCollectionStillAlive] = std::move(value);
+            }
+
+            bool* m_isCollectionStillAlive{};
+            MapT& m_collection;
```

Comment:
We discussed this offline. I would recommend using a double pointer here instead.